### ISSUE TITLE
Debug Console should support find

### ISSUE THREAD

**Issue by gregg-miskelly** [microsoft/vscode#4338]
- VSCode Version: 0.10.10
- OS Version: Ubuntu 14.04

This goes along with a few other debug console feature requests that I see (ex: #3974), but it would also be nice if find worked in the debug console.

Steps to reproduce:
1. Start debugging something, have the console be full of text
2. Hit Ctrl-F to find

Result: Find works in the editor window instead of the debug console

**Comment by sergio-garcia** [comment]
The Ctrl+A also selects all text on the editor windows instead of the debug console.

**Comment by GregRos** [comment]
Is there any status on this? It seems like a really basic feature. Every other editor in existence has it I think. There is even search in the terminal, which I'd think would be harder to implement.

**Comment by kevinsalerno** [comment]
Can we get this please? VSCode is brilliant, but debugging in-editor is 1/10 for this lack of select all and/or find in debug console

**Comment by GregRos** [comment]
@kevinsalerno You can use Select All. Right click and pick "Copy All". However, you still can't use Ctrl+A.

**Comment by ChristosChristofidis** [comment]
But even if you right click copy all you will be missing the lines

**Comment by ncannasse** [comment]
Big up for this, having to search through debug logs requires manual copy/pasting right now, not very optimal...

**Comment by GregRos** [comment]
@ncannasse Exactly my experience. In my project I have a lot of debug logs...

**Comment by welkie** [comment]
Came here when discovering the debugger doesn't support find. Would also like to see this. I find it easier to read in the debugging console than the integrated terminal, so I'd like to not have to redirect output to the integrated terminal if I don't have to.

**Comment by dafeldib** [comment]
Hi, definitely need this added to help with many development tasks.

**Comment by welkie** [comment]
Would like this too. I know some people redirect their debug output to the terminal and work there, but I don't like doing this. I prefer to keep my debug console for my debugging and my terminal for my terminal stuff. Would like to see Find support in the debug console.

**Comment by Mesieu** [comment]
@welkie redirecting the debug output to the terminal sounds like the way to go for now.
Any tip on how to do that?

**Comment by welkie** [comment]
@Mesieu You can use the property `"console": "integratedTerminal"` on the debug launch task to redirect the output to the terminal.

**Comment by MaadhavShah** [comment]
This is honestly essential, idk why it's not implemented yet

**Comment by snapcorp** [comment]
I'd like to add this as the only feature I've ever requested because it's frankly bizarre that it's not there, it's one of those basic things, not a fancy thing, I think.

**Comment by welkie** [comment]
@snapcorp I bet the reason it hasn't been implemented yet is because a lot of the developers I encounter set the debugger to output into the integrated terminal instead. I seem to be the only one I know who chooses to keep it in the internal console. This is probably a Node dev thing, because the popular logging libraries choose to log to stdout instead of the console (which is a good idea of course, it's more performant), but VS Code by default will only show console output in the internal console.

**Comment by PoLoMoTo** [comment]
@welkie Correct me if I'm wrong but I think that may only work with the Node debugger.  I'm using the Java debugger and in my brief research it seems that the Node debugger is the only one that supports outputting to the terminal.

**Comment by yi-ji** [comment]
This function is desperately needed, can't believe it's not been considered...
@welkie thanks but when I try adding `"console": "integratedTerminal"` into `launch.json` it said this field is not allowed

**Comment by RuiffCardoso** [comment]
After adding support for multiple debug consoles, this is the only step that is missing for me to switch to "Debug Console" instead of "Integrated Terminal"

**Comment by Zied-Guesmi** [comment]
I like vscode to the point that I'm using it for spring boot application instead of java IDEs. But this feature is really essential !

**Comment by isidorn** [comment]
From tomorrow's vscode insider the debug console supports highlighting and filtering (just like the explorer for example).
So just click on the debug console and start typing and the matches will get highlighted / filtered.

This is not the exact find behavior, but I believe is a very nice workaround.
Try it out and let us know how it feels.

**Comment by isidorn** [comment]
One of next milestones (probably march) I plan to add a filter functionality to the debug console.
It would behave very similarly to the PROBLEMS view filter. So try that one out and please raise concerns if you have any with regards to taking that approach

@sandy081 it would be cool if I could reuse some code from the PROBLEMS view. If the input box is complex we could potentially turn it into a filter widget or something. Can you also give me some pointers?

**Comment by sandy081** [comment]
Here is the filter action item I used in Problems view.

https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/contrib/markers/electron-browser/markersPanelActions.ts#L118

Take a look and I am open to suggestions. We can also discuss about your requirements and can come up with a common widget.

**Comment by drayfar** [comment]
@isidorn type-to-filter works ok, but it's completely undiscoverable. Just some feedback about that functionality.

**Comment by fel-dev** [comment]
Has the type-to-filter functionality been removed?
Clicking on the debug console does not trigger type-to-filter anymore.

**Comment by isidorn** [comment]
Based on feedback from here https://github.com/microsoft/vscode/issues/68793
We have gone back to the old behavior of clicking anywhere focusing the debug console input.
To focus the tree and type-to-filter we have introduced a new command (ctrl/cmd + F). So you have to press ctrl/cmd + F before filtering.
I should have mentined that in the release notes, but I forgot, sorry.

**Comment by GraemeAllanBryce** [comment]
This is nice.  I agree that it is not exactly discoverable but then that is true of a lot of VS code and it keeps the UI clean.

**Comment by GraemeAllanBryce** [comment]
Is it intentional that the input does not support spaces?  I can filter on "foo" but not on "foo bar".  I can't understand why this limitation has been made.

**Comment by YisraelV** [comment]
@GraemeAllanBryce I think it would be a problem to use space because it's already used to expand nodes in the tree. To see this print a complex object to the console, than expand it using space.

**Comment by YisraelV** [comment]
@isidorn should we consider a find option, instead of or beside a filter?
I think that lots of time you want to see the context of the log i.e. what was logged just before or just after. This is only possible with find. Also this the original issue is about find - so it seems like lots of people wanted find behaviour.

**Comment by isidorn** [comment]
@YisraelV I do agree that find is the ultimate goal here, however since the Tree already supports the filter that might be a "low hanging fruit" solution.
For the find there is more work to be done, but yes that would be the good solution.
As for reusing the find widget, could you please give some pointers since I know you use it in the terminal @Tyriar @rebornix

**Comment by isidorn** [comment]
@YisraelV after discussing this in the standup we feel like the best first approach is to go with the filter. And try adding an input box similar to something what the problem panel has.
You can open a PR, so we start the discussion there on the potential implementation. Thanks!

**Comment by Tyriar** [comment]
@isidorn the terminal and webview use `SimpleFindWidget`, they both sub-class it as `TerminalFindWidget` and `WebviewFindWidget`. The debug console might be able to use the editor's find widget more easily though since it's a monaco instance @rebornix

**Comment by isidorn** [comment]
@Tyriar the debug console is actually not a monaco instance, it is a tree underneath.

**Comment by cvasani** [comment]
> @Mesieu You can use the property `"console": "integratedTerminal"` on the debug launch task to redirect the output to the terminal.

Thanks.

**Comment by devel0** [comment]
> > @Mesieu You can use the property `"console": "integratedTerminal"` on the debug launch task to redirect the output to the terminal.
> 
> Thanks.

however integratedTerminal doesn't work in conjunction with `ServerReadyAction` in launch.json even with insider edition even seems [it should](https://github.com/Microsoft/vscode/issues/70855#issuecomment-475008559).

```json
// Enable launching a web browser when ASP.NET Core starts. For more information: https://aka.ms/VSCode-CS-LaunchJson-WebBrowser
"serverReadyAction": {                
  "action": "debugWithChrome",                                
  "pattern": "^\\s*Now listening on:\\s+(https?://\\S+)"
},
```

**Comment by scullabyte** [comment]
Any update on this? I was able to use `"console": "terminal",` on the debugger, however, it stopped working about a month or so ago. (`integratedTerminal` gets type-checked by VSCode telling me it isn't an option, btw). If anyone knows of any ways to quickly search debug output with the current setup I would love to hear it. I was copy/pasting console output into Sublime Text, until I found this:
https://stackoverflow.com/questions/51268169/how-to-search-the-debug-console-in-vscode
(this is still very far from ideal, as you can't copy/paste or use spaces)

Maybe I'm missing something?

**Comment by ride4sun** [comment]
I am wondering that too. After a long time away from VSCODE  I tried it again and besides that multiple windows are still not fully supported filtering in the debug console falls under deal-breaker for me. I also could not find an extension to add that functionality. Do I miss something ???
This is open longer than 4 years! I am using Flutter btw. I guess back to Android Studio

**Comment by douniwan5788** [comment]
I need "find" not "filter". I want to see the context of the keyword, not the keyword itself!

**Comment by DarkTrick** [comment]
> I need "find" not "filter". I want to see the context of the keyword, not the keyword itself!

yes! 
I have unit tests for my typescript project. However, they are not much worth if more than one test fails, because there is too much output, that can not be searched in.

**Comment by mattwelke** [comment]
Not that it completely solves the problem, but there's a technique I use when I have massive test output and need to search in it. I run the program from the terminal and redirect output to a file. I can open the file in VS Code as the program runs in the terminal and watch my output stream into the file in real time (like a long running Mocha test suite for example). Because it's a file opened in VS Code, I can use all the search features I want to search through the output. When I'm done, I delete the file. If I do it often, I add the file name to .gitignore.

It looks like this:

```
npm test > testout.txt
```

(and then I open the file while that runs)

**Comment by DavidGoldman** [comment]
Are there any plans to do so? Are contributions welcome? Any idea what it would take to change this?

**Comment by burekas7** [comment]
The year is 2022
Is there any progress or update on this issue?

It is so requested and necessary to have the possibility to perform a search in the Console Log, it is strange that this ability is not available until today.
There is only the filter option.
I need to make a find search without filtering.

**Comment by djankey** [comment]
> The year is 2022 Is there any progress or update on this issue?
> 
> It is so requested and necessary to have the possibility to perform a search in the Console Log, it is strange that this ability is not available until today. There is only the filter option. I need to make a find search without filtering.

Unbelievable. It's the end of the year 2022, and this feature is still not implemented.
Even XCode has this :(

![Screenshot 2022-12-30 at 01 26 02](https://user-images.githubusercontent.com/364756/210023443-c1f225b0-ca7e-4898-8dbf-feb77b8fe676.png)

**Comment by bvoq** [comment]
+1
By clicking inside the text of a single log and then pressing Cmd+F immediately, I'm able to get this search view, but it's buggy and doesn't display search results. 
<img width="1124" alt="image" src="https://user-images.githubusercontent.com/3169217/224728910-c7360edf-0b54-44df-bb38-c34f7e646b01.png">

**Comment by roblourens** [comment]
Thanks for pointing that out @bvoq, that's not supposed to show up, and isn't implemented to work properly here yet 😅

**Comment by seryte** [comment]
I need "find" not "filter". I want to see the context of the keyword, not the keyword itself!

**Comment by hacnguyen1412** [comment]
nothing has changed since 2016

**Comment by erabti** [comment]
Any updates?

**Comment by ride4sun** [comment]
It is hard to believe this does not get attention from VS Code. Is this not a low-hanging fruit that would get immediate better user experience? This is what we fix in my project first.

**Comment by jmonroynieto** [comment]
I'm working on a TUI with lots of events that repeat during the rendering process. Yet another example where this filter is an insufficient solution. 

Alternative ui/ux would be to double click to disable filter(temporarily), jump to line and put a highlight on it. Pressing scape takes you back to the filtered results. Since each little result box seems to be a separate element, finding the element itself seems like a viable solution.

**Comment by jmonroynieto** [comment]
I just checked into it, and the list_id is recalculated when filtering. ~sigh

**Comment by Stitch-Taotao** [comment]
OMG,is this feature so complex to implement ？

**Comment by marctan** [comment]
still not implemented wtf?? we need find not filter!

**Comment by gjsjohnmurray** [comment]
@roblourens should this be tagged as debug-console too? It used to be until @isidorn removed that tag on Dec 15, 2020.

**Comment by kyrosle** [comment]
right now is 2023/10/26, vscode still not support `find` in debug console?

**Comment by holland-hng** [comment]
> right now is 2023/10/26, vscode still not support `find` in debug console?
That's right, not only that, they also react with dislikes to questions about this issue

**Comment by bersbersbers** [comment]
> they also react with dislikes to questions about this issue

To "their" defense, this is not "them" - this is me. (And I am not affiliated with Microsoft or VS Code.) 

Since you mentioned it: I routinely dislike comments that contribute nothing substantial in issues with hundreds of subscribers. I do that to give feedback to the respective authors, hoping they get the point without having to trigger notifications for everyone else. Remember that you are essentially emailing hundreds of foreigners across the world as you post here.

Upvote the issue to express your support. However, stay humble and keep in mind that, very likely, your vote is not more important than that of the other 168 voters before you.

**Comment by yasinarik** [comment]
Top of my vscode needed-features list. This is fundamental.

**Comment by kyrosle** [comment]
sorry that, my word my cause some bad dispute, i am a newbie programer, for this topic, i think we can track the debug console event and filiter out the ouput event, and then put them into the webview editor view, this is the project i start to trying : https://github.com/kyrosle/develop-kit, hoping it can provide new ideas

**Comment by kayces** [comment]
It's 2024, and it's astonishingly unbelievable that VS Code still doesn't support a search function in the debugging console. As someone working in client-side development, I find this particularly frustrating. The logs that I need to monitor are constantly getting buried under a deluge of logs from various third-party SDKs. The existing filter functionality is far from sufficient - it only shows the exact matching line, completely stripping away the context provided by preceding and following lines. This makes it virtually impossible to effectively track down and understand the issues at hand. For a modern IDE that's widely used in the development community, this lack of a basic but crucial feature is a significant oversight and a major hindrance to productivity.

**Comment by gjsjohnmurray** [comment]
I just published https://marketplace.visualstudio.com/items?itemName=georgejames.searchable-debug-console that may serve as a workaround.

**Comment by XKMR** [comment]
user @roblourens mentions that such feature exists, but I fail to find it.
https://github.com/microsoft/vscode/issues/169648

**Comment by asapsmc** [comment]
It's now mid-2024, still no implementation available?

**Comment by meganrogge** [comment]
Hi everyone, find is now available in the debug console and can be activated using `ctrl/cmd+alt/option+f`. We have chosen to go with this keybinding for now because it aligns with the tree's find keybinding, which is what we use for this implementation.

ATM, tree find **does not** provide support for highlighting matches in the output. It might be possible to do this when https://github.com/microsoft/vscode/issues/212077 has been implemented. If that is possible and doesn't cause perf issues, then the behavior of find will better align with that in the editor and we can consider switching the default keybinding to match that one, `ctrl/cmd+f`. 

I realize a primary part of this feature request was to have `ctrl/cmd+f` open find in the debug console. For now, you can feel free to manually set this yourself.

We have left the filter widget in the debug console because find doesn't support negative patterns ATM, bolstered by some user feedback https://github.com/microsoft/vscode/pull/225093#issuecomment-2277989162.

**Comment by abhijit-chikane** [comment]
but now the this filter control does not get focus on Ctrl + f in insider release 
![image](https://github.com/user-attachments/assets/8693be57-e297-42e0-8b5c-f5f131eb75a4)

**Comment by meganrogge** [comment]
Yep, it does as it always did before - when you have a debug session active.

**Comment by meganrogge** [comment]
That should work the same as it always have. I'll look into it.

**Comment by meganrogge** [comment]
Thanks for calling this out @abhijit-chikane ! will be fixed in insider's tomorrow #226661
